Gradle regression in tests when a module has a "sidecar" module

### Describe the bug

From Quarkus 3.22.0 to Quarkus 3.23.0 there is a regression in `@QuarkusTest` when a module `A` uses a class from module `B` that reference a class from module `A`. In this situation, a `ClassNotFoundException` is thrown. There is no problem in Quarkus 3.21.4.

Let's say module `A` has a class `Person`:
```java
public record Person(String id) {}
```

And module `B` a `PersonFactory`:
```java
public class PersonFactory {
	public static Person newRandomPerson() {
		return new Person(UUID.randomUUID().toString());
	}
}
```

module `A` imports `B` as a test dependency:
```kotlin
dependencies {
    ...
    testImplementation(project(":B"))
    testImplementation("io.quarkus:quarkus-junit5")
}
```

And module `B` imports `A` as a main dependency:
```kotlin
dependencies {
    implementation(project(":A"))
}
```

If module `A` uses `PersonFactory#newRandomPerson`  in a `@QuarkusTest`, a `ClassNotFoundException` is thrown:
```java
@QuarkusTest
class PersonTest {
	@Test
	void test() {
		PersonFactory.newRandomPerson();
	}
}
```

With the stacktrace being:
```
java.lang.NoClassDefFoundError: com/module/Person
	at com.module.test.PersonFactory.newRandomPerson(PersonFactory.java:9)
	at com.module.PersonTest.test(PersonTest.java:12)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:999)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:847)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ClassNotFoundException: com.module.Person
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:549)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:523)
	... 7 more
```

### How to Reproduce?

Here's a reproducer:
[test-classloader-issue.zip](https://github.com/user-attachments/files/20526481/test-classloader-issue.zip)

1. Go to `PersonFactory` and launch the test
    - A `ClassNotFoundException` should be thrown
2. In `gradle.properties` use the Quarkus version 3.21.4
3. Run the test again
   - The test should pass

### Output of `uname -a` or `ver`

Darwin MacBook-Pro.local 24.5.0 Darwin Kernel Version 24.5.0: Tue Apr 22 19:53:27 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6041 arm64

### Output of `java -version`

openjdk version "21.0.5" 2024-10-15 LTS OpenJDK Runtime Environment Temurin-21.0.5+11 (build 21.0.5+11-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.22.0 to 3.23.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.13 and 8.14

### Additional information

May be related to https://github.com/quarkusio/quarkus/issues/48157, but I am not sure.

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusApplicationModelTask.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusApplicationModelTask.java
@@ -31,6 +31,7 @@ import org.gradle.api.artifacts.ModuleVersionIdentifier;
 import org.gradle.api.artifacts.ResolvableDependencies;
 import org.gradle.api.artifacts.component.ComponentArtifactIdentifier;
 import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
 import org.gradle.api.artifacts.result.DependencyResult;
 import org.gradle.api.artifacts.result.ResolvedArtifactResult;
 import org.gradle.api.artifacts.result.ResolvedComponentResult;
@@ -186,7 +187,7 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
         }
     }
 
-    private static void collectDependencies(QuarkusResolvedClasspath classpath, ApplicationModelBuilder modelBuilder,
+    private void collectDependencies(QuarkusResolvedClasspath classpath, ApplicationModelBuilder modelBuilder,
             WorkspaceModule.Mutable wsModule, ProjectDescriptor projectDescriptor) {
         final Map<ComponentIdentifier, List<QuarkusResolvedArtifact>> artifacts = classpath
                 .resolvedArtifactsByComponentIdentifier();
@@ -195,8 +196,11 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
         final Set<ModuleVersionIdentifier> processedModules = new HashSet<>();
         classpath.getRoot().get().getDependencies().forEach(d -> {
             if (d instanceof ResolvedDependencyResult resolved) {
-                final byte flags = (byte) (COLLECT_TOP_EXTENSION_RUNTIME_NODES | COLLECT_DIRECT_DEPS
-                        | COLLECT_RELOADABLE_MODULES);
+                byte flags = (byte) (COLLECT_TOP_EXTENSION_RUNTIME_NODES | COLLECT_DIRECT_DEPS);
+                final LaunchMode launchMode = getLaunchMode().get();
+                if (!launchMode.equals(LaunchMode.NORMAL)) {
+                    flags |= COLLECT_RELOADABLE_MODULES;
+                }
                 collectDependencies(resolved, modelBuilder, artifacts, wsModule, alreadyCollectedFiles,
                         processedModules, flags, projectDescriptor);
             }
@@ -311,8 +315,15 @@ public abstract class QuarkusApplicationModelTask extends DefaultTask {
                 newFlags = clearFlag(newFlags, COLLECT_RELOADABLE_MODULES);
             }
             if (isFlagOn(flags, COLLECT_RELOADABLE_MODULES)) {
-                if (projectModule != null) {
-                    depBuilder.setFlags(DependencyFlags.RELOADABLE);
+                // Checking whether current dependency is a project module is a temporary workaround,
+                // that is required while projectModule for project dependencies is null (current
+                // deficiency of this task).
+                // That's why we set the workspace module flag explicitly via setWorkspaceModule().
+                // Once we have projectModule set for project dependencies, we can remove this workaround.
+                final boolean isProjectDependency = resolvedDependency.getSelected()
+                        .getId() instanceof ProjectComponentIdentifier;
+                if (projectModule != null || isProjectDependency) {
+                    depBuilder.setReloadable().setWorkspaceModule();
                     modelBuilder.addReloadableWorkspaceModule(artifactKey);
                 } else {
                     newFlags = clearFlag(newFlags, COLLECT_RELOADABLE_MODULES);

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/gradle.properties
@@ -1,0 +1,2 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/build.gradle
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/build.gradle
@@ -1,0 +1,23 @@
+plugins {
+    id 'java'
+    id 'io.quarkus'
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex 'io.quarkus.*'
+            includeGroup 'org.hibernate.orm'
+        }
+    }
+    mavenCentral()
+    gradlePluginPortal()
+}
+
+dependencies {
+    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
+    implementation "io.quarkus:quarkus-arc"
+
+    testImplementation project(":test-support")
+    testImplementation "io.quarkus:quarkus-junit5"
+}

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/src/main/java/com/module/Person.java
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/src/main/java/com/module/Person.java
@@ -1,0 +1,4 @@
+package com.module;
+
+public record Person(String id) {
+}

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/src/test/java/com/module/PersonTest.java
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/my-module/src/test/java/com/module/PersonTest.java
@@ -1,0 +1,16 @@
+package com.module;
+
+import org.junit.jupiter.api.Test;
+
+import com.module.test.PersonFactory;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+class PersonTest {
+
+    @Test
+    void test() {
+        PersonFactory.newRandomPerson();
+    }
+}

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/settings.gradle
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/settings.gradle
@@ -1,0 +1,19 @@
+pluginManagement {
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex 'io.quarkus.*'
+                includeGroup 'org.hibernate.orm'
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+      id 'io.quarkus' version "${quarkusPluginVersion}"
+    }
+}
+rootProject.name = 'test-with-sidecar-module'
+
+include "test-support"
+include "my-module"

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/test-support/build.gradle
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/test-support/build.gradle
@@ -1,0 +1,13 @@
+plugins {
+    id 'java-library'
+}
+
+repositories {
+    mavenLocal()
+    mavenCentral()
+    gradlePluginPortal()
+}
+
+dependencies {
+    implementation project(":my-module")
+}

--- a/integration-tests/gradle/src/main/resources/test-with-sidecar-module/test-support/src/main/java/com/module/test/PersonFactory.java
+++ b/integration-tests/gradle/src/main/resources/test-with-sidecar-module/test-support/src/main/java/com/module/test/PersonFactory.java
@@ -1,0 +1,12 @@
+package com.module.test;
+
+import java.util.UUID;
+
+import com.module.Person;
+
+public class PersonFactory {
+
+    public static Person newRandomPerson() {
+        return new Person(UUID.randomUUID().toString());
+    }
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/TestWithSidecarModule.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/TestWithSidecarModule.java
@@ -1,0 +1,18 @@
+package io.quarkus.gradle;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+
+import org.junit.jupiter.api.Test;
+
+// Reproduces https://github.com/quarkusio/quarkus/issues/48159
+public class TestWithSidecarModule extends QuarkusGradleWrapperTestBase {
+
+    @Test
+    public void test() throws Exception {
+        final File projectDir = getProjectDir("test-with-sidecar-module");
+        final BuildResult build = runGradleWrapper(projectDir, "clean", ":my-module:test");
+        assertThat(BuildResult.isSuccessful(build.getTasks().get(":my-module:test"))).isTrue();
+    }
+}

